rewrite extended eventhandlers

commy2 opened the pull request:
- adds: RPT message for every class that overwrites XEH (including the addon name)
- adds: PreStart XEH event on game start
- adds: XEH are cached on game start (less compiling on first mission start needed when using -world=empty)
- adds: Easier to make XEH compatible addon that doesn't require CBA by using new EventHandlers sub classes
- fixes: multiple case sensitivity issues (every manual input is case insensitive now)
- fixes: `scope = 0;` unpredictable under certain conditions
- fixes: OO-like overriding of event handlers partially broken
- fixes: GetInMan firing when entering AND leaving a vehicle
- fixes: GetOutMan never firing?
- fixes: https://github.com/CBATeam/CBA_A3/issues/42
- closes: https://github.com/CBATeam/CBA_A3/pull/188
- removes: enable auto XEH, no longer needed

Killswitch00 commented on addons/xeh/fnc_startFallbackLoop.sqf:
```diff
@@ -0,0 +1,42 @@
+    };
+}, 0.1, []] call CBA_fnc_addPerFrameHandler;
+
+GVAR(fallbackRunning) = true;
```
Move to line 25, directly after `if (GVAR(fallbackRunning)) ...` to avoid running more than once

commy2 commented:
made some more changes:
- added a macro for a reoccurring, long if-expression for better code readability
- support monitor now ignores classes with scope 0/private. (Prevents triggering the fall back loop for `rhs_su25_testbed` and `rhs_mi24v_testbed`. Those can't be spawned anyway.)
- Always show incompatible classes on game start. (Really useful information that would otherwise be lost when starting __AllInArmaTerrainpack__.)
- Moved some Functions to CfgFunctions, so `CBA_fnc_addClassEventHandler` won't fail when used on BI preInit in an addon without using `requiredAddons[] = {"cba_xeh"};`
- Fixes: Adding a "Init" during a mission (after object initialization) would call the init-event on all XEH-incompatible units as if they just spawned in.

Glowbal commented:
HandleDamage does not behave as in Vanilla - it should use the last returned value from handleDamage that is not `nil`.

commy2 commented:
I tried applying a return value to the config handleDamage variant and the event does work (it didn't in A2 ?), but the return value is ignored.

jokoho48 commented on addons/xeh_a3/CfgEventHandlers.hpp:
```diff
@@ -1,249 +0,0 @@
-        class Land_Communication_F /* : House_F */ {
-                SLX_BIS = "[(_this select 0)] execVM ""\A3\Structures_F\Ind\Transmitter_Tower\Scripts\tower_ruins.sqf""";
-        };
-};
```
i think this is still Required

commy2 replied:
It's no longer overwritten, since I use a sub class (new in v1.54) for the XEH events.

jokoho48 replied:
ok sorry missed that :+1:

Killswitch00 commented:
*Crosses fingers*

PabstMirror commented on addons/common/XEH_preInit.sqf:
```diff
@@ -18,8 +18,6 @@ SCRIPT(XEH_preInit);
  * and will hopefully ensure forward compatibility with future ArmA II patches.
  */
 
-if (isNil "CBA_FUNC_RECOMPILE") then { CBA_FUNC_RECOMPILE = CACHE_DIS(functions) };
```
Used in https://github.com/CBATeam/CBA_A3/blob/master/addons/common/init_functionsModule.sqf#L13 (now nil)

esteldunedain commented:
Congrats!
